### PR TITLE
[breaking] type-safe `Maybe.all` implementation

### LINKED ISSUE #87: Type-safe `Maybe.tuple` and `Maybe.all`
[Using][playground] the variadic tuple types newly introduced [in TypeScript 4.0][ts-4.0], reimplement `Maybe.all` and `Maybe.tuple` in a type-safe manner.

Note that this requires dropping support for TS versions prior to 4.0—thus the work done in #86.

[playground]: https://www.typescriptlang.org/play?#code/MYGwhgzhAEBSCuEAuAeAKgPmgbwFDQOgBMBTUMAJxOiTAHMAuaAcgWWYG59DgB7AO2QV4wJLwoAKKmCICQAT2gA3MCHgkmaAJQ4Avrn25yUaADleSABYBLfnXRY8hYmXBUa9Js3NXbdTtwEfIJIwqLiEjrY+oZI8gAO1ACyYPIARiQO0AC8cIiomNAAPmYWNnYOXLhxidAAqvwA7hRg8YlEWblo0CQAHkgk-EQwKemZtgBmJBT1WAD89dCaVTXUaPDxIJndfQNDMACCFC3yKKMZKPD8ANb8vI38GE85gdA7-YPD0ADaAHT-k2m9QAutBXs4FudMtgfgBpaC2aDXEjyXgTEFMBrNVrtFB1b6w4FYXQYcGEZa4aoJagATRybw2WxQ3yhKCEfgwABpoKz+PAALYZCgYIlcVbQABa9PWm0yRxOZ1SF3ZdiwJV5AqFTwwHCAA

[ts-4.0]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types

### PR DESCRIPTION
**Pull request by chriskrycho**
- [breaking] Reimplement `Maybe.all` using variadic tuple types.
- Deprecate `Maybe.tuple` as extraneous now that `Maybe.all` can handle both tuples and arrays.
- Update dev version of TS to 4.1.3
- [breaking] In preparation for implementing #87, remove support for versions prior to TypeScript 4.0, which introduced the variadic tuples type feature required to implement `Maybe.all` and `Maybe.tuple` safely.

Resolves #87

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,48 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
...
+### Changed :boom:
+
+- `Maybe.all` is now type-safe and can correctly handle both array and tuple types. To support this change, it now accepts arrays or tuples directly, and the
+variadic/spread arguments to `all` have been replaced with taking an array or
+tuple directly. While `tuple` is unchanged, it is also deprecated (see below).
+
+    **Before:**
+
+    ```ts
+    import Maybe, { all, just, nothing, tuple } from 'true-myth/maybe';
```
Comment (bmakuh) [inline CHANGELOG.md:18]: You're not actually using `tuple` in either of these examples; did you mean to?

File: src/maybe.ts [hunk 2]
```diff
@@ -1596,58 +1596,36 @@ export function last<T>(array: Array<T | null | undefined>): Maybe<T> {
...
   ```ts
   import Maybe from 'true-myth/maybe';
 
   let valid = [Maybe.just(2), Maybe.just('three')];
-  Maybe.all(...valid); // => Just([2, 'three']);
+  let allJust = Maybe.all(valid); // => Just([2, 'three']);
 
-  let invalid = [Maybe.just(2), Maybe.nothing<string>()];
-  Maybe.all(...invalid); // => Nothing
+  let invalid: Invalid = [Maybe.just(2), Maybe.nothing<string>()];
```
Comment (bmakuh) [inline src/maybe.ts:1623]: Where is this `: Invalid` type coming from?

File: CHANGELOG.md (hunk 1 above)
Comment (chriskrycho) [reply CHANGELOG.md:18]: Haha, nope! Will update removing it. Thank you!

File: src/maybe.ts (hunk 2 above)
Comment (chriskrycho) [reply src/maybe.ts]: My imagination, as far as I can tell. 😂 Thanks for catching it; I've removed it.

**Pull request merged**